File: python/CardGameEx/Ex04.py
# Ex04.py

from ch.aplu.jgamegrid import Location
from ch.aplu.jcardgame import Deck, CardGame, StackLayout, RowLayout
from ch.aplu.jcardgame.Hand import SortType
from enum import enum

def dealingOut():
   cg.setStatusText("Dealing Out...")
   hands = deck.dealingOut(nbPlayers, 0)  # All cards in talon
   talon = hands[nbPlayers]

   for i in range(nbPlayers):
      hands[i].setView(cg, RowLayout(handLocations[i], 300, 90 * i))
      hands[i].draw()
      hands[i].setTouchEnabled(True)

   talon.setView(cg, StackLayout(handLocations[nbPlayers]))
   talon.draw()

   while not talon.isEmpty() and not cg.isDisposed():
      for i in range(nbPlayers):
         top = talon.getLast()
         talon.transfer(top, hands[i], False)
         hands[i].sort(SortType.RANKPRIORITY, True)

# ------------------------ main --------------------------------------------
Suit = enum("SPADES", "HEARTS", "DIAMONDS", "CLUBS")
Rank = enum("ACE", "KING", "QUEEN", "JACK", "TEN")
deck = Deck(Suit.values(), Rank.values(), "cover");

nbPlayers = 4
handLocations = [Location(300, 525), Location(75, 300), Location(300, 75), Location(525, 300), Location(300, 300)]
  
cg = CardGame(600, 600, 30)
dealingOut()
cg.setStatusText("Dealing Out...done");

File: cardgame_tutorial/Ex04.java
// Ex04.java

import ch.aplu.jcardgame.*;
import ch.aplu.jgamegrid.*;

public class Ex04 extends CardGame
{
  public enum Suit
  {
    SPADES, HEARTS, DIAMONDS, CLUBS
  }

  public enum Rank
  {
    ACE, KING, QUEEN, JACK, TEN
  }
  //
  private final Deck deck =
    new Deck(Suit.values(), Rank.values(), "cover");
  private final int nbPlayers = 4;
  private final Location[] handLocations =
  {
    new Location(300, 525),
    new Location(75, 300),
    new Location(300, 75),
    new Location(525, 300),
    new Location(300, 300)
  };
  private Hand[] hands;
  private Hand talon;

  public Ex04()
  {
    super(600, 600, 30);
    dealingOut();
    setStatusText("Dealing Out...done.");
  }

  private void dealingOut()
  {
    setStatusText("Dealing Out...");

    hands = deck.dealingOut(nbPlayers, 0);  // All cards in talon
    talon = hands[nbPlayers];

    for (int i = 0; i < nbPlayers; i++)
    {
      hands[i].setView(this, new RowLayout(handLocations[i], 300, 90 * i));
      hands[i].draw();
      hands[i].setTouchEnabled(true);
    }

    talon.setView(this, new StackLayout(handLocations[nbPlayers]));
    talon.draw();

    while (!talon.isEmpty())
    {
      for (int i = 0; i < nbPlayers; i++)
      {
        Card top = talon.getLast();
        talon.transfer(top, hands[i], false); // Don't show to avoid flickering
        hands[i].sort(Hand.SortType.RANKPRIORITY, true);
      }
    }
  }

  public static void main(String[] args)
  {
    new Ex04();
  }
}
